src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,419 @@
+
+    /**
+     * An internal mutex used rather than synchronizing on `this`. Lazily created.
+     * If `null`    , we have not entered a mutex section yet
+     * If TOMBSTONE , we do not need synchronization anymore (isSet() && isError() = true)
+     * if instanceof Throwable , records a previous error and, we do not need synchronization anymore
+     * otherwise    , a distinct synchronization object
+     */
+    private Object mutex;
+
+    /**
+     * If non-null, holds a set value
+     * If `null`  , may be unset or hold a set `null` value
+     */
+    @Stable
+    private V value;
+
+    /**
+     * If StableUtil.NOT_SET  , a value is not set
+     * If StableUtil.NON_NULL , a non-null value is set
+     * If StableUtil.NULL     , a `null` value is set
+     */
+    @Stable
+    private int state;
+
+    /**
+     * Indicates if a supplier is currently being invoked. Used to
+     * detect circular supplier invocations.
+     */
+    @Stable private boolean supplying;
+
+    private StableValueImpl() {}
+
+    @ForceInline
+    @Override
+    public boolean isSet() {
+        int s;
+        return (s = state) == NON_NULL || s == NULL ||
+                (s = stateVolatile()) == NON_NULL || s == NULL;
+    }
+
+    @ForceInline
+    @Override
+    public boolean isError() {
+        return state == ERROR || stateVolatile() == ERROR;
+    }
+
+    @ForceInline
+    @Override
+    public V orThrow() {
+        // Optimistically try plain semantics first
+        final V v = value;
+        if (v != null) {
+            // If we happen to see a non-null value under
+            // plain semantics, we know a value is set.
+            return v;
+        }
+        if (state == NULL) {
+            // If we happen to see a state value of NULL under
+            // plain semantics, we know a value is set to `null`.
+            return null;
+        }
+        // Now, fall back to volatile semantics.
+        return orThrowVolatile();
+
+/*        // This is intentionally an old switch statement as it generates
+        // more compact byte code.
+        switch (state) {
+            case UNSET:    { throw StableUtil.notSet(); }
+            case NULL:     { return null; }
+            case NON_NULL: { return value; }
+            case ERROR:    { throw StableUtil.error(this); }
+            case DUMMY:    { throw shouldNotReachHere(); }
+        }
+        throw shouldNotReachHere();*/
+
+    }
+
+    @DontInline // Slow-path taken at most once per thread if set
+    private V orThrowVolatile() {
+        // This is intentionally an old switch statement as it generates
+        // more compact byte code.
+        switch (stateVolatile()) {
+            case UNSET:    { throw StableUtil.notSet(); }
+            case NULL:     { return null; }
+            case NON_NULL: { return valueVolatile(); }
+            case ERROR:    { throw StableUtil.error(this); }
+            case DUMMY:    { throw shouldNotReachHere(); }
```

Comment:
Redundant branch?

Reply:
The idea here is to have the most likely value in the middle... Not sure if that motivates the added complexity though.

Reply:
Is there any refernce on how/why the middle entry in a tableswitch instruction is the fastest?

Reply:
It is only in a _lookupswitch_ that this becomes relevant. The above code will generate a *tableswitch* so I think it is safe to simplify the code and remove the DUMMY.

```
  private V orThrowVolatile();
    descriptor: ()Ljava/lang/Object;
    flags: (0x0002) ACC_PRIVATE
    Code:
      stack=1, locals=1, args_size=1
         0: aload_0
         1: invokevirtual #15                 // Method stateVolatile:()I
         4: tableswitch   { // 0 to 4
                       0: 40
                       1: 44
                       2: 46
                       3: 51
                       4: 56
                 default: 60
            }
            ...
```